### PR TITLE
[#21] Task10 - Maximum product of three numbers

### LINKED ISSUE #21: Task10 - Maximum product of three numbers
Return the maximum product of three numbers in an array, so [-8, -9, 1, 2, 7] outputs 504 (which is -8*-9*7). Float are allowed.

### PR DESCRIPTION
**Pull request by MaraEfimova**
First, I checked if the input string was valid.
I sorted the array, took the first two elements and the last three. Since arrays with 3 and 4 elements are valid, which means that the previous condition can collect the same elements, I used the .unic method.
After that, I wrote a function that sequentially multiplies all the numbers in the array and chose the largest number.
**Resolves #21**

### CODE REVIEW

File: task10.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Return the maximum product of three numbers in an array, so [-8, -9, 1, 2, 7] outputs 504 (which is -8*-9*7)
+
+def task10(arr)
+  raise ArgumentError, "Not a valid format. Need an array" if arr.class != Array
+  raise ArgumentError, "Array is empty. This is not valid" if arr.empty?
+  raise ArgumentError, "Not a valid format. Need an array with more then two integer or float numbers inside" if arr.length < 3 
+  arr.each do |number|
+    raise ArgumentError, "Not a valid format. Need an array with more then two integer or float numbers inside" if number.class != Integer && number.class != Float
```
Comment (kaloyan-yordanov) [inline task10.rb:8]: @MaraEfimova I believe this is not the same error as the one above, if it is - then combine the conditionals and do not repeat it :wink:

File: task10.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  arr.sort!
+  arr = [arr[0], arr[1], arr[-1], arr[-2], arr[-3]].uniq.sort
```
Comment (kaloyan-yordanov) [inline task10.rb]: @MaraEfimova Could you please remind me if the requirement says that there should not be repeted numbers (e.g. `[5, 5, 5]`)?

File: task10.rb (hunk 1 above)
Comment (kaloyan-yordanov) [inline task10.rb]: @MaraEfimova I believe this is not the same error as the one above, if it is - then combine the conditionals and do not repeat it :wink:

File: task10.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  max_product = arr[0]*arr[1]*arr[2]
+  n = arr.length
+  i = 0
+
+  while i < n-2 do
```
Comment (kaloyan-yordanov) [inline task10.rb]: @MaraEfimova Did you check if Ruby has a built-in method that could create the combinations for you?

**Review by kaloyan-yordanov — CHANGES_REQUESTED** [review]

File: task10.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+def task10(arr)
+  raise ArgumentError, "Not a valid format. Need an array" if arr.class != Array
+  raise ArgumentError, "Array is empty. This is not valid" if arr.empty?
+  raise ArgumentError, "Not a valid format. Need an array with more then two numbers inside" if arr.length < 3 
+  arr.each do |number|
+    raise ArgumentError, "Not a valid format. Elements must be Integer or Float numbers." if number.class != Integer && number.class != Float
+  end
+
+  if arr.length >= 5
```
Comment (kaloyan-yordanov) [inline task10.rb]: @MaraEfimova Would you remind me why is this sorting needed here?

File: task10.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Return the maximum product of three numbers in an array, so [-8, -9, 1, 2, 7] outputs 504 (which is -8*-9*7)
+
+def task10(arr)
+  raise ArgumentError, "Not a valid format. Need an array" if arr.class != Array
+  raise ArgumentError, "Array is empty. This is not valid" if arr.empty?
+  raise ArgumentError, "Not a valid format. Need an array with more then two numbers inside" if arr.length < 3 
+  arr.each do |number|
```
Comment (kaloyan-yordanov) [inline task10.rb]: @MaraEfimova Not a mistake but you're not sure if this is a `number`, so until you are sure - it is still just an `element` :wink:

**Review by kaloyan-yordanov — CHANGES_REQUESTED** [review]
@MaraEfimova In general works fine but still needs to be polished a bit.

File: task10.rb (hunk 4 above)
Comment (MaraEfimova) [reply task10.rb]: If there are many elements in the array, this will help to avoid time-consuming creation of combinations.

Comment (MaraEfimova) [reply task10.rb]: So I sorted them and then take only 5 (two biggest from negative numbers and three positive numbers)

Comment (kaloyan-yordanov) [reply task10.rb]: @MaraEfimova Sorting of long arrays might sometimes be heavier than just calculating a few combinations more, so I would say it really depends. And here you're actually using a "hack". But OK, let's leave it like it is :+1:

**Review by kaloyan-yordanov — APPROVED** [review]
@MaraEfimova Ready to go, please squash and merge :+1:

**Pull request merged**